### PR TITLE
posts: Add 'GSoC23 Final Report' post

### PR DESCRIPTION
**Pull request by davidbtadokoro**

### CODE REVIEW

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
+---
+layout: post
+title: GSoC23 Final Report
+date: 2023-08-26 08:00:00 -0300
+categories: [gsoc23]
+tags: [gsoc23, kw, kw patch-hub, lore, linux]
+---
+
+My GSoC23 journey, which I introduced in a [previous post](), is almost over. It really doesn't feel like 16 weeks have passed, but I can say that, in this period, I have learned a lot and grown as a developer.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: I guess you missed the link in the `previous post` part.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+This feature is named `kw patch-hub`, and this blog post is a "final report" of my GSoC23 contributions.
+
+<br>
+
+## **Non-related Contributions**
+
+---
+
+This first section is to describe my contributions to kw that are not directly related to my GSoC23 project. Nonetheless, these were important in their own context and made me more in sync with kw coding style, its contribution model, and, most importantly, with my mentors and people around the project (which I found most invaluable).
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: How about: This first section describes my contributions to kw that are not directly related to the patch-hub feature but were part of my GSoC.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+### **Adding Support for Native Zsh Completions**
+
+Before GSoC23, I already was contributing to kw and the first big one (both in scope and in number of commits) that I did was adding support for native Zsh completions. Without getting into too much detail, each Shell, say, Bash, can provide command completions. In other words, it's the well-known behavior of hitting `TAB` and waiting for the Shell to either complete the command you are typing or show possible completions.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: >  I already was contributing to kw and the first big one (both in scope and in number of commits) that I did was adding support for native Zsh completions.

You made those contributions in the context of the GSoC application. How about something like this:

> I contributed meaningful changes to kw project during the application period. My first significant contribution (both in scope and number of commits) was adding support for native Zsh completions.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+These completions are Shell-dependent, and kw only had native support for Bash completions. Zsh completions were adapted from the native Bash completions, but this "emulation" didn't work and resulted in broken completions for Zsh. This was a waste, as the Zsh completion system (although really complicated to work on) provided deeper features than Bash, like highlighting options shown, coupling documentation with options shown, and more.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: `although really complicated to work on` drop this part.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+During February I worked on bringing kw support for native Zsh completions. I described this in further detail in an [earlier post](), but you can see the full Pull Request by clicking [here](https://github.com/kworkflow/kworkflow/pull/773). To illustrate, below is a demo of the results achieved.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: Add post link.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+![Kw Zsh Completion]({{site.url}}/images/gifs/kw-zsh-completion.gif)
+
+### **Introducing SQLite3 to kw**
+
+From before the Community Bonding Period began until halfway through it (from mid-April to mid-May), I worked on introducing the Database Management System (DBMS) SQLite3 to kw. This was a long-awaited addition for the kw community, as it would improve the project's scalability and allow the collection of statistics.
+
+I discussed this in further detail in an [earlier post]() and the full Pull Request representing this contribution can be accessed [here](https://github.com/kworkflow/kworkflow/pull/836).
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: add link

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+3. Capabilities of replying patchsets in the public mailing lists with Reviewed-by, Tested-by, and inline reviews.
+
+> I use the term *patchset* instead of *patch*, because a patchset is a logical set of patches pertaining to the same context, while a patch is any individual change sent as a message. For reviewing, considering chunks of related changes instead of individual changes makes more sense. Just think of reviewing a Pull Request in its whole context, versus reviewing the commits of this PR independently.
+{: .prompt-info }
+
+### **First Cycle: Understanding the Problem and Building the Core**
+
+Since my proposal, my better understanding of the problem at hand along with my interactions with my mentors, made me realize that the most important deliverable was to provide a reliable UI to lore.
+
+Not that we planned on having discrete development cycles, but looking in hindsight, I can divide my work on `kw patch-hub` into 3 development cycles. The first cycle was related to experimenting and understanding the problem and building the feature's core.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: Do you mean `Note`?

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+From an organizational perspective, we documented every starting requisite in issues and added them to a GitHub Kanbam board. I wish I could provide a link to this board, but its visibility is private, so below is a print of it, just for illustration purposes.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: I didn't know the kanbam was private; sorry for that. I made It public.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+From the code perspective, I focused on understanding what was already done, what needed to be done, what needed to change, and built the core of `kw patch-hub`. About two years ago, my mentors [Rodrigo Siqueira](https://siqueira.tech/) and [Melissa Wen](https://melissawen.github.io/) implemented what we can call the "predecessor" of `kw patch-hub` named `kw upstream-patches-ui`. This was mostly a prototype that validated the feature, but that laid the foundation needed for my project.
+
+At the end of this cycle, `kw patch-hub` started to look functional and the feature software architecture was somewhat solidified (although, as we will see in a moment, kind of messy). At that moment, the feature was still named `kw upstream-patches-ui` and looked like this:
+
+![kw patch-hub First Cycle]({{site.url}}/images/gifs/kw_patch_hub_first_cycle.gif)
+
+From mid-May up until mid-June, those were my contributions in the form of Pull Requests, in chronological order of merge:
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: Maybe highlight the number of commits per PR.

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+
+### **Second Cycle: Refactoring**
+
+As mentioned, `kw patch-hub` had a core implemented, however, the feature badly needed refactoring.
+
+At this point, the feature was implemented across three files: 2 library files (`src/lib/lore.sh` and `src/lib/dialog_ui.sh`) and one that represented the feature itself (`src/upstream_patches_ui.sh`). The Model-View-Controller was softly implemented in a way that `src/lib/lore.sh` was the Model, `src/lib/dialog_ui.sh` was the View, and `src/upstream_patches_ui.sh` was the Controller.
+
+##### Refactoring the Controller
+
+I described in a [previous post]() the Finite-State Machine computation model used to implement `kw patch-hub` Controller, but the thing was that for each new state added, `src/upstream_patches_ui.sh` grew uncontrollably. At one moment, the file was more than 500 lines in size with functions that didn't follow a logical order, which made it harder and harder to scroll to the desired line each time an addition was made. To exemplify the need for refactoring on this Controller front, there was a `switch-case` with more than 100 lines.
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md]: Add link

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+1. Optimize fetch time. In the demo GIF above you can see that loading times are not good.
+2. Fix parsing of patchsets. In the demo GIF above you can see some patchsets metadata malformatted/incorrect.
+3. Add an 'Apply' action for patchsets. 
+4. Add a 'Build' action for patchsets.
+5. Add a 'Deploy' action for patchsets.
+6. Add query based on string. In other words, integrate a more refined search of lore archives on the feature.
+7. Allow users to reply patchsets with 'Reviewed-by', 'Tested-by', and with inline reviews.
+8. Improve feature UX.
+9. Refine feature fixing bugs.
+
```
Comment (rodrigosiqueira) [inline _posts/2023-08-26-gsoc23-final-report.md:328]: Let everyone know that you will keep working on the project until the end of this year since this is also your capstone project.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Please keep the lines short to make it easy to review the new version.

**Comment by davidbtadokoro** [comment]
Thank you so much for the review @rodrigosiqueira.

Sorry for the delay. I updated the post on my site yesterday, but forgot to update the PR here. The links to the posts that you commented were left blank because these posts aren't already in kw's blog. I added links to the posts on my site and will change them to point to kw's blog if they get merged into it.

Thanks again for the review!

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+[kw patch-hub: Add reliable fetch of latest patchsets from mailing list](https://github.com/kworkflow/kworkflow/pull/889)
+
+<br>
+
+## **Next Steps**
+
+---
+
+Considering the deliverables, I implemented number 1 and part of number 2. More
```
Comment (melissawen) [inline _posts/2023-08-26-gsoc23-final-report.md]: I didn't get what number 1 and 2 are here... what are the deliverables? Also, I really think you've delivered a huge work, but the way you expressed it here doesn't make it clear... it just the tone, but I'd prefer something like:

```
[As a result of my GSoC project, `kw patch-hub` can be used as a reliable] UI to the lore archives
and provides some other functionalities like bookmarking patchsets, downloading
applicable patchsets (to a default or custom directory), and managing the feature's
settings through the feature itself.
```

File: _posts/2023-08-26-gsoc23-final-report.md [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+and provides some other functionalities like bookmarking patchsets, downloading
+applicable patchsets (to a default or custom directory), and managing the feature's
+settings through the feature itself.
+
+It's important to note that `kw patch-hub` has become an integral part of my Capstone
+Project, so I'll keep updating the feature until the end of this year, and probably
+further than that.
+
+Here is a list of practically all steps needed for `kw patch-hub` to be considered
+"complete", not necessarily in the order of importance:
```
Comment (melissawen) [inline _posts/2023-08-26-gsoc23-final-report.md]: I'd prefer another perspective here. Think incrementally: `kw patch-hub` exists and can be used. As any active project, it will always require new features, improvements and maintenance, therefore you've mapped the next steps to enhance the tool and cover developers/user needs better. ;)

File: _posts/2023-08-26-gsoc23-final-report.md (hunk 12 above)
Comment (davidbtadokoro) [reply _posts/2023-08-26-gsoc23-final-report.md]: You are right. That's a much better way to say it. Also, thanks a lot for the compliment in my work :)

File: _posts/2023-08-26-gsoc23-final-report.md (hunk 13 above)
Comment (davidbtadokoro) [reply _posts/2023-08-26-gsoc23-final-report.md]: For sure, I agree. This notion of "all steps needed" and "complete" isn't real. Thanks 😄

**Comment by rodrigosiqueira** [comment]
Merged! Thanks a lot.
I'm looking forward to the other posts.
